sweagent/investigations/google_drive.py

```diff
@@ -0,0 +1,237 @@
+    service = get_google_drive_service()
+    try:
+        # Make sure we can access the folder first.
+        service.files().get(fileId=folder_id).execute()
+    except Exception as error:
+        print(f"⚠ Failed to access folder with ID at {get_google_drive_folder_href(folder_id)} ⚠.")
+        print("Listing all shared folders the service has access to:\n\n")
+        list_shared_folders()
+        # raise error
+        return []
+
+    # Find matching files.
+    query = f"'{folder_id}' in parents and {query}"
+    results = service.files().list(q=query, fields="files(id, name, size)").execute()
+    items = results.get("files", [])
+    print(f"Downloading {len(items)} files matching query '{query}'")
+
+    if not items:
+        print(f"Could not find any log files in folder {folder_id}.")
+        return []
+
+    downloaded_files = []
+    skipped = 0
+    for item in tqdm(items, desc="Processing files", unit="file"):
+        file_name = item["name"]
+        local_file_path = get_local_file_path(file_name)
+        file_id = item["id"]
+        file_size = int(item["size"])
+        skipped += download_drive_file(local_file_path, file_id, file_size)
+        downloaded_files.append(local_file_path)
+
+    print(f"Finished downloading ({skipped} skipped).")
+
+    return list(set(downloaded_files))
+
+def list_shared_folders() -> list[str]:
+    """
+    List all shared folders that the service account has access to.
+    """
+    service = get_google_drive_service()
+    try:
+        query = "mimeType = 'application/vnd.google-apps.folder' and sharedWithMe"
+        results = service.files().list(q=query, fields="files(id, name)").execute()
+        items = results.get("files", [])
+
+        if not items:
+            print("No shared folders found.")
+            return []
+
+        print("Shared folders:")
+        for item in items:
+            print(f"Folder ID: {item['id']}, Name: {item['name']}")
+
+        return [item["id"] for item in items]
+
+    except HttpError as error:
+        print(f"An error occurred while listing shared folders: {str(error)}")
+        return []
+
+def upload_file(dest_parent_folder_id: str, relative_dest_path: list[str], local_file_path: str):
+    service = get_google_drive_service()
+    
+    # Get or create the destination folder
+    dest_folder_id = get_or_create_drive_folder(dest_parent_folder_id, relative_dest_path)
+    
+    file_metadata = {
+        'name': os.path.basename(local_file_path),
+        'parents': [dest_folder_id]
+    }
+    
+    media = MediaFileUpload(local_file_path, resumable=True)
+    
+    try:
+        file = service.files().create(body=file_metadata, media_body=media, fields='id').execute()
+        return file.get('id')
+    except HttpError as error:
+        print(f'An error occurred: {error}')
+        return None
+
+def upload_directory(dest_parent_folder_id: str, relative_dest_path: list[str], local_directory_path: str):
```

Comment:
Here you use `directory`, but the other places all use the term `folder` instead. Can we stick with one please?

* Google drive uses the term `folder`.
* I also personally prefer it, because it won't inevitably get abbreviated with `dir` as soon as people feel a little lazy 🤷

Reply:
nod.  I was using directory to refer to the local side / folder to the google drive side.  It was getting pretty confusing while writing the code.

now that it's done/working I can flip them over though.